target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-tr4400-v2.dts

```diff
@@ -0,0 +1,387 @@
+				read-only;
+			};
+			partition@54e0000 {
+				label = "0:SBL3_1";
+				reg = <0x54e0000 0x0280000>;
+				read-only;
+			};
+			partition@5760000 {
+				label = "0:DDRCONFIG_1";
+				reg = <0x5760000 0x0120000>;
+				read-only;
+			};
+			partition@5880000 {
+				label = "0:SSD_1";
+				reg = <0x5880000 0x0120000>;
+				read-only;
+			};
+			partition@59a0000 {
+				label = "0:TZ_1";
+				reg = <0x59a0000 0x0280000>;
+				read-only;
+			};
+			partition@5c20000 {
+				label = "0:RPM_1";
+				reg = <0x5c20000 0x0280000>;
+				read-only;
+			};
+			partition@5ea0000 {
+				label = "0:BOOTCONFIG1";
+				reg = <0x5ea0000 0x0060000>;
+				read-only;
+			};
+			partition@5f00000 {
+				label = "0:APPSBL_1";
+				reg = <0x5f00000 0x0500000>;
+				read-only;
+			};
+			stock_partition@6400000 {
+				label = "stock_rootfs_1";
+				reg = <0x6400000 0x4000000>;
+			};
+			stock_partition@a400000 {
+				label = "stock_fw_env";
+				reg = <0xa400000 0x0100000>;
+			};
+			stock_partition@a500000 {
+				label = "stock_config";
+				reg = <0xa500000 0x0800000>;
+			};
+			stock_partition@ad00000 {
+				label = "stock_PKI";
+				reg = <0xad00000 0x0200000>;
+			};
+			stock_partition@af00000 {
+				label = "stock_scfgmgr";
+				reg = <0xaf00000 0x0100000>;
+			};
+
+			fw_env: partition@6400000 {
+				label = "fw_env";
+				reg = <0x6400000 0x0100000>;
+			};
+			partition@6500000 {
+				label = "ubi";
+				reg = <0x6500000 0x9b00000>;
+			};
+			partition@1340000 {
+				label = "extra";
+				reg = <0x1340000 0x4000000>;
+			};
+		};
+	};
+};
+
+&mdio0 {
+	status = "okay";
+
+	pinctrl-0 = <&mdio0_pins>;
+	pinctrl-names = "default";
+
+	phy0: ethernet-phy@0 {
+		reg = <0x0>;
+		qca,ar8327-initvals = <
+			0x00004 0x7600000   /* PAD0_MODE */
+			0x00008 0x1000000   /* PAD5_MODE */
+			0x0000c 0x80        /* PAD6_MODE */
+			0x000e4 0xaa545     /* MAC_POWER_SEL */
+			0x000e0 0xc74164de  /* SGMII_CTRL */
+			0x0007c 0x4e        /* PORT0_STATUS */
+			0x00094 0x4e        /* PORT6_STATUS */
+			>;
+	};
+
+	phy4: ethernet-phy@4 {
+		reg = <4>;
+	};
+
+	phy7: ethernet-phy@7 {
+		reg = <7>;
+	};
+};
+
+&gmac0 {
+	status = "okay";
+	phy-mode = "rgmii";
+	qcom,id = <0>;
+	phy-handle = <&phy4>;
+
+	nvmem-cells = <&fw_env_macaddr_4_unknown>;
+	nvmem-cell-names = "mac-address";
+
+	pinctrl-0 = <&rgmii2_pins>;
+	pinctrl-names = "default";
+
+	fixed-link {
+		speed = <1000>;
+		full-duplex;
+	};
+};
+
+&gmac1 {
+	status = "okay";
+	phy-mode = "sgmii";
+	qcom,id = <1>;
+
+	nvmem-cells = <&fw_env_macaddr_0_eth1_lan>;
+	nvmem-cell-names = "mac-address";
+
+	fixed-link {
+		speed = <1000>;
+		full-duplex;
+	};
+};
+
+&gmac3 {
+	status = "okay";
+	phy-mode = "sgmii";
+	qcom,id = <3>;
+	phy-handle = <&phy7>;
+
+	nvmem-cells = <&fw_env_macaddr_1_eth2_wan>;
+	nvmem-cell-names = "mac-address";
+};
+
+&adm_dma {
+	status = "okay";
+};
+
+&usb3_0 {
+	status = "okay";
+	clocks = <&gcc USB30_1_MASTER_CLK>;
+};
+
+&usb3_1 {
+	status = "okay";
+	clocks = <&gcc USB30_0_MASTER_CLK>;
+};
```

Comment:
Your device has only one USB port, so I guess you need only one node?

Reply:
yes, this is true, and more serious. IDK if i have the time though. note that the stock dtb has these two entries.

Reply:
```
root@OpenWrt:/# lsusb 
Bus 002 Device 001: ID 1d6b:0003 Linux 5.10.109 xhci-hcd xHCI Host Controller
Bus 004 Device 001: ID 1d6b:0003 Linux 5.10.109 xhci-hcd xHCI Host Controller
Bus 001 Device 002: ID 0a12:0001  BT DONGLE10
Bus 001 Device 001: ID 1d6b:0002 Linux 5.10.109 xhci-hcd xHCI Host Controller
Bus 003 Device 001: ID 1d6b:0002 Linux 5.10.109 xhci-hcd xHCI Host Controller
root@OpenWrt:/# lspci 
0000:00:00.0 PCI bridge: Qualcomm Device 0101
0000:01:00.0 Network controller: Qualcomm Atheros QCA9984 802.11ac Wave 2 Wireless Network Adapter
0001:00:00.0 PCI bridge: Qualcomm Device 0101
0001:01:00.0 Network controller: Qualcomm Atheros QCA9980/9990 802.11ac Wireless Network Adapter
```

@mans0n no way that this could be a usb (minus some kind of phy)?
https://openwrt.org/inbox/toh/arris/tr4400_v2?do=#jtag

Reply:
seems to have 2 hci pairs:
```
root@OpenWrt:/# dmesg | egrep -i '(usb|(o|u|e|x)hci)'
[    0.066633] usbcore: registered new interface driver usbfs
[    0.066721] usbcore: registered new interface driver hub
[    0.066793] usbcore: registered new device driver usb
[    6.860938] ehci_hcd: USB 2.0 'Enhanced' Host Controller (EHCI) Driver
[    6.882766] ehci-fsl: Freescale EHCI Host controller driver
[    6.883447] ehci-platform: EHCI generic platform driver
[    6.900598] ohci_hcd: USB 1.1 'Open' Host Controller (OHCI) Driver
[    6.901087] ohci-platform: OHCI generic platform driver
[    6.907661] dwc3-qcom 110f8800.usb3: IRQ hs_phy_irq not found
[    6.910867] dwc3-qcom 110f8800.usb3: IRQ dp_hs_phy_irq not found
[    6.916689] dwc3-qcom 110f8800.usb3: IRQ dm_hs_phy_irq not found
[    6.922852] dwc3-qcom 110f8800.usb3: IRQ ss_phy_irq not found
[    6.930647] dwc3-qcom 100f8800.usb3: IRQ hs_phy_irq not found
[    6.934397] dwc3-qcom 100f8800.usb3: IRQ dp_hs_phy_irq not found
[    6.940201] dwc3-qcom 100f8800.usb3: IRQ dm_hs_phy_irq not found
[    6.946203] dwc3-qcom 100f8800.usb3: IRQ ss_phy_irq not found
[    7.093340] xhci-hcd xhci-hcd.0.auto: xHCI Host Controller
[    7.093389] xhci-hcd xhci-hcd.0.auto: new USB bus registered, assigned bus number 1
[    7.097876] xhci-hcd xhci-hcd.0.auto: hcc params 0x0228f065 hci version 0x100 quirks 0x0000000002010010
[    7.105400] xhci-hcd xhci-hcd.0.auto: irq 51, io mem 0x11000000
[    7.115602] hub 1-0:1.0: USB hub found
[    7.124562] xhci-hcd xhci-hcd.0.auto: xHCI Host Controller
[    7.128452] xhci-hcd xhci-hcd.0.auto: new USB bus registered, assigned bus number 2
[    7.133824] xhci-hcd xhci-hcd.0.auto: Host supports USB 3.0 SuperSpeed
[    7.148011] usb usb2: We don't know the algorithms for LPM for this host, disabling LPM.
[    7.148351] hub 2-0:1.0: USB hub found
[    7.160057] xhci-hcd xhci-hcd.1.auto: xHCI Host Controller
[    7.163784] xhci-hcd xhci-hcd.1.auto: new USB bus registered, assigned bus number 3
[    7.169221] xhci-hcd xhci-hcd.1.auto: hcc params 0x0228f065 hci version 0x100 quirks 0x0000000002010010
[    7.176793] xhci-hcd xhci-hcd.1.auto: irq 52, io mem 0x10000000
[    7.187249] hub 3-0:1.0: USB hub found
[    7.196048] xhci-hcd xhci-hcd.1.auto: xHCI Host Controller
[    7.199902] xhci-hcd xhci-hcd.1.auto: new USB bus registered, assigned bus number 4
[    7.205113] xhci-hcd xhci-hcd.1.auto: Host supports USB 3.0 SuperSpeed
[    7.212869] usb usb4: We don't know the algorithms for LPM for this host, disabling LPM.
[    7.220016] hub 4-0:1.0: USB hub found
[    7.569534] usb 1-1: new full-speed USB device number 2 using xhci-hcd
```

but there is something strange with the clock defs. they are swapped?
```
  usb3_0: usb3@110f8800 {
   compatible = "qcom,dwc3", "syscon";
   #address-cells = <1>;
   #size-cells = <1>;
   reg = <0x110f8800 0x8000>;
   clocks = <&gcc 267>;
   clock-names = "core";
...

&usb3_0 {
 status = "okay";
 clocks = <&gcc 268>;
};
```
```
  usb3_1: usb3@100f8800 {
   compatible = "qcom,dwc3", "syscon";
   #address-cells = <1>;
   #size-cells = <1>;
   reg = <0x100f8800 0x8000>;
   clocks = <&gcc 268>;
   clock-names = "core";
...

&usb3_1 {
 status = "okay";
 clocks = <&gcc 267>;
};
```

i'll try leaving clock defs out.

Reply:
works without the clock defs.

Reply:
these drivers are loaded but seem unused, as the hw is xhci.
```
[    7.450651] ehci_hcd: USB 2.0 'Enhanced' Host Controller (EHCI) Driver
[    7.471760] ehci-fsl: Freescale EHCI Host controller driver
[    7.472709] ehci-platform: EHCI generic platform driver
[    7.482572] ohci_hcd: USB 1.1 'Open' Host Controller (OHCI) Driver
[    7.483386] ohci-platform: OHCI generic platform driver
```

Reply:
pci stuff is not related to usb.
Did you inserted the BT dongle yourself or is it built-in one? If the usb port is not used by on-board module, you can safely disable it as it is not used and it cannot be used.

Reply:
@mans0n

without both these:
```
&usb3_X {
	status = "okay";
};
```
defs (X = 0 & 1) the USB does not work.

with only &usb3_0..... bootloop!! kernel hangs, and i suppose watchdog gets angry
trying only &usb3_1 now...

> Did you insert the BT dongle yourself

yes i put it there to test if USB responds.

Reply:
@mans0n ok the usb...

- this works:
```
&usb3_0 {
	status = "okay";
};

&usb3_1 {
	status = "okay";
};
```

- with both defs absent kernel boots but obviously no usb.
- with either def alonge, the kernel hangs and bootloops.

my hypothesis: the clocks are swapped. if only one block is enabled, the kernel clocks the wrong one and talks to the dead, and hell breaks loose.

if true, the dtsi would be wrong here. i'll try swapping the clocks from the dts but i dunno if that overrides the dtsi.

for now seems both blocks must be enabled.

Reply:
this alone kinda works (no bootloop):
```
&usb3_0 {
	status = "okay";
	clocks = <&gcc USB30_1_MASTER_CLK>;
};
```
but then when i remove the dongle:
```
[  114.207167] usb 1-1: USB disconnect, device number 2
[  124.209588] xhci-hcd xhci-hcd.0.auto: Abort failed to stop command ring: -110
[  124.241641] xhci-hcd xhci-hcd.0.auto: Host halt failed, -110
[  124.241675] xhci-hcd xhci-hcd.0.auto: xHCI host controller not responding, assume dead
[  124.246392] xhci-hcd xhci-hcd.0.auto: HC died; cleaning up
[  124.254392] xhci-hcd xhci-hcd.0.auto: Timeout while waiting for configure endpoint command
```
and stays dead

something is wrong outside of my dts.

Reply:
can it be that the clocks of the two USB3 host controllers got swapped?
The names kinda indicate that `USB30_1_MASTER_CLK` would belong to `usb3_1` and `USB30_0_MASTER_CLK` to `usb3_0`.

Reply:
> can it be that the clocks of the two USB3 host controllers got swapped? The names kinda indicate that `USB30_1_MASTER_CLK` would belong to `usb3_1` and `USB30_0_MASTER_CLK` to `usb3_0`.

they seem to be swapped in the dtsi. please read the whole section.

Reply:
ok...

the block in use is usb3_0 (not _1). however the clocks seem to be swapped somewhere in the included files. adding a clock in the dts does something but does not fully work around the problem: maybe the kernel turns both clocks on and off for each block.

in any case, i'll keep both USB blocks enabled as that seems to be the only effective workaround.

well congrats to us! we just hurt the planet. we wasted amp-hours in building openwrt tens of times, with the prospect of maybe clock off a block and save a few micro-amps in a handful of devices that would never add up to compensate that down the road... no to mention time :)

Reply:
Thanks for the detailed report. I don't understand why the clocks are swapped, but I think it's okay to leave both of the usb nodes turned on for now.
And one thing that just came to mind is that the usb phys (`ss_phy_X` and `hs_phy_X`) might need to be swapped too...or not.

Reply:
the facts:
1) as the dtsi defines it (no clock added in dts):
   a) if you enable either single block: the kernel hangs on boot, then bootloops.
   b) if you enable both blocks: the USB port seems to work.
2) if you add the "wrong" clock in the dts (in addition to the "right" one in the dts):
   a) if you enable only usb3_0: the USB port works on boot. but as soon as you unplug a USB device, the xhci block stops responding and the driver quickly declares it dead. this might happen as a side effect of usb3_1 also declaring both clocks: the kernel might turn off usb3_1 again, which also causes usb3_0 to die.
   b) if you enable only usb3_1: the USB port work does not work.
   c) if you enable both blocks: the USB port seems to work.

so workable setups seem to be 1b) and 2c). i chose 2c) because:
- no explicit clocks is clearly not working as expected.
- the explicit "wrong" clocks calls attention to this problem. <-- main reason
- somebody that knows more chose this setup:
  - 1b) could have issues that 2c) solves, IDK.
- double clock defs seems not worse than wrong clock defs.

> I think it's okay to leave both of the usb nodes turned on for now
changing the dtsi affects lots of devices so i don't have a choice.

besides, i don't even know their origin: are the dtsi's provided by qualcomm?

Reply:
@mans0n 
and this is fishy:
```
  hs_phy_0: hs_phy_0 {
   reg = <0x110f8800 0x30>;
  ss_phy_0: ss_phy_0 {
   reg = <0x110f8830 0x30>;
  usb3_0: usb3@110f8800 {
   reg = <0x110f8800 0x8000>;
   dwc3_0: dwc3@11000000 {
    reg = <0x11000000 0xcd00>;

  hs_phy_1: hs_phy_1 {
   reg = <0x100f8800 0x30>;
  ss_phy_1: ss_phy_1 {
   reg = <0x100f8830 0x30>;
  usb3_1: usb3@100f8800 {
   reg = <0x100f8800 0x8000>;
   dwc3_1: dwc3@10000000 {
    reg = <0x10000000 0xcd00>;
```
so:
- all usb 0: 0x110xxxxx
- all usb 1: 0x100xxxxx

weird naming. coupled to the clock issue, smells rotten.

if there is a mixup at all -not silicon errata-, all things mostly static could be implicated:
- clocks
- enables
- resets
- interrupts even

this potential rabbit hole is clearly outside the scope of this port. but do you want to dig deeper? i enabled CCACHE so i'm faring better here, #facepalm. i'd still want to prioritize backporting if possible.